In [8]:
# import pandas as pd
# import matplotlib.pyplot as plt
# import seaborn as sns
# import plotly.graph_objects as go
# import numpy as np

In [2]:
inventory = pd.read_parquet('C:/Users/neddy/Desktop/스파르타 코딩클럽/최종 프로젝트/데이터셋 및 코딩 파일/inventory_optimized.parquet')

In [3]:
log = pd.read_parquet('C:/Users/neddy/Desktop/스파르타 코딩클럽/최종 프로젝트/데이터셋 및 코딩 파일/real_log_for_spartan_optimized.parquet')
log

CustomerID ActionType  GoodsCode   Price       AdID     CPC  \
0             74377       view       1484  108330        NaN     NaN   
1             74377   purchase       1484  108330        NaN     NaN   
2             74377       view       1063   14370        NaN     NaN   
3             74377   purchase       1063   14370        NaN     NaN   
4             74377   ad_click       1203   85290  ad_google  175.38   
...             ...        ...        ...     ...        ...     ...   
3096926      454879       view       1250  100420        NaN     NaN   
3096927      454879       view       1250  100420        NaN     NaN   
3096928      454879       view       1250  100420        NaN     NaN   
3096929      454879     basket       1250  100420        NaN     NaN   
3096930      454879   purchase       1250  100420        NaN     NaN   

                         Timestamp  
0       2022-02-17 19:07:19.511967  
1       2022-03-20 04:26:25.511967  
2       2022-03-02 04:43:47.511967  
3       2022-03-22 00:42:08.511967  
4       2022-04-16 22:19:28.511967  
...                            ...  
3096926 2022-05-27 08:18:40.554370  
3096927 2022-04-25 00:41:09.554370  
3096928 2022-06-14 03:44:21.554370  
3096929 2022-05-13 11:09:41.554370  
3096930 2022-06-19 11:04:44.554370  

[3096931 rows x 7 columns]

In [9]:
pd.read_parquet('C:/Users/neddy/Desktop/스파르타 코딩클럽/최종 프로젝트/데이터셋 및 코딩 파일/RFM 최종/purchase_for_product_df.parquet')

CustomerID                  Timestamp  GoodsCode   Price  \
0            74377 2022-03-20 04:26:25.511967       1484  108330   
1           206834 2022-04-21 16:17:55.511967       1484  108330   
2            92485 2022-04-08 16:22:13.511967       1484  108330   
3           272797 2022-04-12 23:32:38.511967       1484  108330   
4            54595 2022-04-22 23:14:36.511967       1484  108330   
...            ...                        ...        ...     ...   
458767      388625 2022-05-08 02:47:04.211784       1275   25910   
458768      503377 2022-04-07 08:21:32.326065       1275   25910   
458769      290404 2022-05-14 17:55:37.600689       1275   25910   
458770      840458 2022-04-25 23:50:27.925560       1275   25910   
458771      896032 2022-03-26 11:05:32.292336       1275   25910   

             Category  cluster_num  
0         Electronics            2  
1         Electronics            2  
2         Electronics            2  
3         Electronics            2  
4         Electronics            0  
...               ...          ...  
458767  Home & Garden            4  
458768  Home & Garden            2  
458769  Home & Garden            4  
458770  Home & Garden            2  
458771  Home & Garden            1  

[458772 rows x 6 columns]

In [4]:
rfm_cluster = pd.read_parquet('C:/Users/neddy/Desktop/스파르타 코딩클럽/최종 프로젝트/데이터셋 및 코딩 파일/RFM 최종/rfm_df.parquet')
rfm_cluster = rfm_cluster.drop('cluster_5', axis = 1).\
    rename(columns = {'cluster_5_weight' : 'cluster', 'recency':'R', 'frequency':'F', 'monetary':'M'})
rfm_cluster

R  F       M  cluster
CustomerID                         
6           318  3   83890        4
8           378  3  146050        2
13          356  3   78560        4
14          367  5   96680        4
17          340  5  213980        1
...         ... ..     ...      ...
899973      337  1   18930        0
899987      351  2   27320        0
899992      370  3   84900        4
899993      368  3  142880        2
899999      333  5   68690        4

[176448 rows x 4 columns]

In [5]:
cluster_log_merge = rfm_cluster.reset_index().merge(log, on = "CustomerID")
cluster_log_purchase = cluster_log_merge[cluster_log_merge['ActionType']=='purchase']
cluster_log_purchase

CustomerID    R  F       M  cluster ActionType  GoodsCode  Price  \
6                 6  318  3   83890        4   purchase       1437  35290   
8                 6  318  3   83890        4   purchase       1047   6670   
12                6  318  3   83890        4   purchase       1486  41930   
16                8  378  3  146050        2   purchase       1372  24930   
18                8  378  3  146050        2   purchase       1013  16210   
...             ...  ... ..     ...      ...        ...        ...    ...   
2864377      899999  333  5   68690        4   purchase       1368   6660   
2864380      899999  333  5   68690        4   purchase       1368   6660   
2864401      899999  333  5   68690        4   purchase       1059  21230   
2864403      899999  333  5   68690        4   purchase       1059  21230   
2864405      899999  333  5   68690        4   purchase       1034  12910   

        AdID  CPC                  Timestamp  
6        NaN  NaN 2022-04-26 21:46:00.166991  
8        NaN  NaN 2022-06-02 08:53:44.166991  
12       NaN  NaN 2022-07-05 04:40:30.166991  
16       NaN  NaN 2022-03-08 03:17:01.156038  
18       NaN  NaN 2022-04-04 03:41:09.156038  
...      ...  ...                        ...  
2864377  NaN  NaN 2022-03-18 23:44:47.722389  
2864380  NaN  NaN 2022-03-21 01:52:58.722389  
2864401  NaN  NaN 2022-04-29 20:42:41.722389  
2864403  NaN  NaN 2022-04-04 08:26:03.722389  
2864405  NaN  NaN 2022-06-20 03:40:43.722389  

[458772 rows x 11 columns]

In [30]:
rfm_cluster.groupby('cluster').count()

R      F      M
cluster                     
0        86124  86124  86124
1        15870  15870  15870
2        30916  30916  30916
3          653    653    653
4        42885  42885  42885

In [6]:
rfm_means = rfm_cluster.groupby('cluster').agg({
    'R' : 'mean',
    'F' : 'mean',
    'M' : 'mean'})
rfm_means = rfm_means.assign(total_mean = rfm_means.sum(axis = 1)-(rfm_means['R']*2)).sort_values(by = 'total_mean')
rfm_means

R         F              M     total_mean
cluster                                                    
0        383.346001  1.422298   34903.843760   34521.920057
3        165.042879  1.555896   49799.418070   49635.931087
4        361.922327  3.622502   75044.356302   74686.056477
2        368.517176  2.779241  153649.653254  153283.915319
1        354.548960  5.922369  236105.347826  235756.721235

In [9]:
cluster_log_purchase['cluster_plus'] = np.where((cluster_log_purchase['cluster'] == 1) & \
    (cluster_log_purchase['M'] > cluster_log_purchase.query('cluster == 1')['M'].quantile(0.9)), 'vip', cluster_log_purchase['cluster'])

C:\Users\neddy\AppData\Local\Temp\ipykernel_3608\125668179.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cluster_log_purchase['cluster_plus'] = np.where((cluster_log_purchase['cluster'] == 1) & \


In [15]:
rfm_means = cluster_log_purchase.groupby('cluster_plus').agg({'R':'mean', 'F':'mean', 'M':'mean'})
rfm_means = rfm_means.assign(sumed_means = rfm_means['F']+rfm_means['M']-rfm_means['R'])
rfm_means.sort_values(by = 'sumed_means')

R         F              M    sumed_means
cluster_plus                                                    
0             381.099058  1.596486   37325.861756   36946.359185
3             163.253937  2.470472   76161.998031   76001.214567
4             361.357371  3.804720   78139.067145   77781.514493
2             367.196536  3.035264  157738.394376  157374.233104
1             354.066048  6.133704  218899.675494  218551.743149
vip           351.586933  7.838476  420426.384337  420082.635880

# 고객 정보 요약

### 클러스터별 요약
##### -평균치 : R 372.14, F 2.6 , M 83617.15
##### -중앙값 : R 370,    F 2,    M 60175
##### -*R은 신규 고객인 3번을 제외하고는 평균이 모두 비슷. 덧붙여 신규 고객 제외하곤 최근 200일 이내 구매한 고객 없음
##### VIP : 우수고객 중 상위 10%
##### 1번(15870명) : 우수고객 => R은 평균이나 M 두드러지게 높으며 F도 6회에 육박 => 상위 10%의 니즈를
##### 2번(30916명) : 중요고객 => R과 F 는 평균보다 약간 좋은 편. 돈을 평균보다 많이 씀.
##### 4번(42885명) : 일반고객 => R과 F 는 평균보다 높으나 돈을 평균보다 적게 씀
##### 3번(653명)   : 신규 고객 => F도 M도 낮지만, 비교적 최근에 구매함(전체 5개월 차이)
##### 0번(86124명) : R, F, M 모두 낮음 => 

In [53]:
print('0번 클러스터 전체 주문일 평균' , cluster_log_purchase[cluster_log_purchase['cluster']==0]['Timestamp'].mean())
print('1번 클러스터 전체 주문일 평균' , cluster_log_purchase[cluster_log_purchase['cluster']==1]['Timestamp'].mean())
print('2번 클러스터 전체 주문일 평균' , cluster_log_purchase[cluster_log_purchase['cluster']==2]['Timestamp'].mean())
print('신규고객 3번 클러스터 전체 주문일 평균' , cluster_log_purchase[cluster_log_purchase['cluster']==3]['Timestamp'].mean())
print('4번 클러스터 전체 주문일 평균' , cluster_log_purchase[cluster_log_purchase['cluster']==4]['Timestamp'].mean())

0번 클러스터 전체 주문일 평균 2022-04-22 10:50:36.758601984
1번 클러스터 전체 주문일 평균 2022-04-18 14:11:47.442083840
2번 클러스터 전체 주문일 평균 2022-04-19 22:48:05.865053696
신규고객 3번 클러스터 전체 주문일 평균 2022-09-13 09:01:17.018804736
4번 클러스터 전체 주문일 평균 2022-04-19 21:18:26.093009664


In [17]:
cluster_statistics = cluster_log_purchase.groupby('cluster_plus').agg(
    n_of_people = ('CustomerID', 'nunique'),
    total_expenditure = ('Price', 'sum'),
    avg_purchase_time = ('Timestamp', 'mean')
).reset_index()

#수 비율
cluster_statistics = cluster_statistics.assign(
    number_percentage = cluster_statistics['n_of_people']/cluster_statistics['n_of_people'].sum())
cluster_statistics['number_percentage'] = cluster_statistics['number_percentage'].round(2)

#지출 비율
cluster_statistics = cluster_statistics.assign(
    expenditure_percentage = cluster_statistics['total_expenditure']/cluster_statistics['total_expenditure'].sum())
cluster_statistics['expenditure_percentage'] = cluster_statistics['expenditure_percentage'].round(2)


cluster_statistics = cluster_statistics.assign(
    payment_per_person = cluster_statistics['total_expenditure']/cluster_statistics['n_of_people']
)
cluster_statistics['payment_per_person'] = cluster_statistics['payment_per_person'].round(2)

cluster_statistics

cluster_plus  n_of_people  total_expenditure             avg_purchase_time  \
0            0        86124         3006058640 2022-04-22 10:50:36.758601984   
1            1        14553         3201414050 2022-04-18 15:52:47.438879232   
2            2        30916         4750232680 2022-04-19 22:48:05.865053952   
3            3          653           32519020 2022-09-13 09:01:17.018804736   
4            4        42885         3218277220 2022-04-19 21:18:26.093009920   
5          vip         1317          545577820 2022-04-17 23:02:42.312381440   

   number_percentage  expenditure_percentage  payment_per_person  
0               0.49                    0.20            34903.84  
1               0.08                    0.22           219983.10  
2               0.18                    0.32           153649.65  
3               0.00                    0.00            49799.42  
4               0.24                    0.22            75044.36  
5               0.01                    0.04           414258.03

In [ ]:
plt.figure(figsize=(4, 4))
plt.pie(cluster_statistics['number_percentage'], labels=cluster_statistics['cluster_plus'], autopct='%1.1f%%', startangle=140)
plt.title('customer percentage by cluster')
plt.axis('equal')  # 원 그래프를 원형으로 보이도록 설정
plt.show()

plt.figure(figsize=(4, 4))
plt.pie(cluster_statistics['expenditure_percentage'], labels=cluster_statistics['cluster_plus'], autopct='%1.1f%%', startangle=140)
plt.title('expenditure percentage by cluster')
plt.axis('equal')  # 원 그래프를 원형으로 보이도록 설정
plt.show()

In [18]:
fig2 = go.Figure(data=[go.Pie(labels=cluster_statistics['cluster_plus'], values=cluster_statistics['number_percentage'])])
fig2.update_layout(title='클러스터별 고객 수 비중')
fig2.show()



fig1 = go.Figure(data=[go.Pie(labels=cluster_statistics['cluster_plus'], values=cluster_statistics['expenditure_percentage'])])
fig1.update_layout(title='클러스터별 매출 비중')
fig1.show()


fig3 = go.Figure(data=[go.Bar(x=cluster_statistics['cluster_plus'], y=cluster_statistics['payment_per_person'])])
fig3.update_layout(title= '클러스터별 인당 지출')
fig3.show()

### 클러스터별 매출 주요 품목 뜯어보기
##### 전체 대상 카테고리별 매출
##### 클러스터별 카테고리 매출 및 비중

In [19]:
cluster_category = cluster_log_purchase.merge(inventory[['GoodsCode', 'Category']], on = 'GoodsCode')


In [20]:
cluster_category

CustomerID    R  F       M  cluster ActionType  GoodsCode  Price AdID  \
0                6  318  3   83890        4   purchase       1437  35290  NaN   
1              378  370  2   45820        0   purchase       1437  35290  NaN   
2              940  348  2   55240        0   purchase       1437  35290  NaN   
3             1057  339  2   55240        0   purchase       1437  35290  NaN   
4             2400  357  2   65590        0   purchase       1437  35290  NaN   
...            ...  ... ..     ...      ...        ...        ...    ...  ...   
458767      891284  378  2   44250        0   purchase       1083  23020  NaN   
458768      893893  457  1   23020        0   purchase       1083  23020  NaN   
458769      894416  343  3   72230        4   purchase       1083  23020  NaN   
458770      896248  362  4  103650        4   purchase       1083  23020  NaN   
458771      897875  361  3   71940        4   purchase       1083  23020  NaN   

        CPC                  Timestamp cluster_plus         Category  
0       NaN 2022-04-26 21:46:00.166991            4    Home & Garden  
1       NaN 2022-04-10 22:24:43.794648            0    Home & Garden  
2       NaN 2022-06-05 08:43:44.052427            0    Home & Garden  
3       NaN 2022-06-14 14:10:37.789614            0    Home & Garden  
4       NaN 2022-04-20 16:06:07.511967            0    Home & Garden  
...     ...                        ...          ...              ...  
458767  NaN 2022-05-06 19:06:15.707955            0  Health & Beauty  
458768  NaN 2022-02-16 14:09:42.038511            0  Health & Beauty  
458769  NaN 2022-05-09 21:46:40.961569            4  Health & Beauty  
458770  NaN 2022-03-19 04:58:31.267931            4  Health & Beauty  
458771  NaN 2022-05-23 23:25:29.525882            4  Health & Beauty  

[458772 rows x 13 columns]

In [21]:
category_by_whole_cluster = cluster_category.groupby('Category').agg(
    count = ('GoodsCode', 'count'),
    total_price = ('Price', 'sum')
    ).reset_index()
category_by_whole_cluster = category_by_whole_cluster.assign(n_ratio = (category_by_whole_cluster['count']/category_by_whole_cluster['count'].sum()).round(2))
category_by_whole_cluster = category_by_whole_cluster.assign(t_price_ratio = (category_by_whole_cluster['total_price']/category_by_whole_cluster['total_price'].sum()).round(2))
category_by_whole_cluster = category_by_whole_cluster.assign(price_per_unit = (category_by_whole_cluster['total_price']/category_by_whole_cluster['count']).round(2))
category_by_whole_cluster = category_by_whole_cluster[['Category', 'count','n_ratio', 'total_price', 'price_per_unit', 't_price_ratio']]
category_by_whole_cluster

Category   count  n_ratio  total_price  price_per_unit  \
0            Books  102507     0.22   1307432410        12754.57   
1         Clothing   87169     0.19   1565272870        17956.76   
2      Electronics   99617     0.22   8223498070        82551.15   
3  Health & Beauty   85816     0.19   1350092820        15732.41   
4    Home & Garden   83663     0.18   2307783260        27584.28   

   t_price_ratio  
0           0.09  
1           0.11  
2           0.56  
3           0.09  
4           0.16

In [73]:
category_by_cluster = cluster_category.groupby(['cluster_plus', 'Category']).agg(
    Category = ('Category', 'first'),
    Count = ('CustomerID', 'count'),
    Total_Price = ('Price', 'sum')
)
cluster_counts = category_by_cluster.groupby('cluster_plus')['Count'].sum().reset_index().rename(columns = {'Count':'summed_count'})
category_by_cluster = category_by_cluster.merge(cluster_counts, on = 'cluster_plus')
category_by_cluster = category_by_cluster.assign(n_ratio_in_cluster = (category_by_cluster['Count']/category_by_cluster['summed_count']).round(2))

cluster_price = category_by_cluster.groupby('cluster_plus')['Total_Price'].sum().reset_index().rename(columns = {'Total_Price' : 'T_Price_by_Cluster'})
category_by_cluster = category_by_cluster.merge(cluster_price, on = 'cluster_plus')
category_by_cluster = category_by_cluster.assign(price_ratio_in_cluster = (category_by_cluster['Total_Price']/category_by_cluster['T_Price_by_Cluster']).round(2))
category_by_cluster = category_by_cluster.drop(['summed_count'], axis = 1)
category_by_cluster = category_by_cluster.merge(cluster_statistics[['cluster_plus', 'payment_per_person']], on = 'cluster_plus')


In [75]:
category_by_cluster.sort_values(by = ['payment_per_person', 'price_ratio_in_cluster'], ascending = False)

cluster_plus         Category  Count  Total_Price  n_ratio_in_cluster  \
27          vip      Electronics   4474    448877600                0.48   
29          vip    Home & Garden   1323     38930650                0.14   
26          vip         Clothing   1220     23062740                0.13   
25          vip            Books   1247     16100110                0.13   
28          vip  Health & Beauty   1134     18606720                0.12   
7             1      Electronics  23938   2062481140                0.28   
9             1    Home & Garden  15052    426552430                0.18   
6             1         Clothing  14677    270224210                0.17   
5             1            Books  16625    214850260                0.20   
8             1  Health & Beauty  14298    227306010                0.17   
12            2      Electronics  40034   3857089490                0.47   
14            2    Home & Garden  12320    360725830                0.14   
11            2         Clothing  11044    205175790                0.13   
13            2  Health & Beauty  10577    171323590                0.12   
10            2            Books  11948    155917980                0.14   
24            4    Home & Garden  30828    820685970                0.20   
22            4      Electronics  13895    730123070                0.09   
21            4         Clothing  34180    602157530                0.22   
23            4  Health & Beauty  34397    534386090                0.22   
20            4            Books  42051    530924560                0.27   
17            3      Electronics    226     18127600                0.22   
19            3    Home & Garden    175      4940270                0.17   
16            3         Clothing    194      3473220                0.19   
18            3  Health & Beauty    197      3157730                0.19   
15            3            Books    224      2820200                0.22   
2             0      Electronics  17050   1106799170                0.14   
4             0    Home & Garden  23965    655948110                0.20   
1             0         Clothing  25854    461179380                0.21   
0             0            Books  30412    386819300                0.25   
3             0  Health & Beauty  25213    395312680                0.21   

    T_Price_by_Cluster  price_ratio_in_cluster  payment_per_person  
27           545577820                    0.82           414258.03  
29           545577820                    0.07           414258.03  
26           545577820                    0.04           414258.03  
25           545577820                    0.03           414258.03  
28           545577820                    0.03           414258.03  
7           3201414050                    0.64           219983.10  
9           3201414050                    0.13           219983.10  
6           3201414050                    0.08           219983.10  
5           3201414050                    0.07           219983.10  
8           3201414050                    0.07           219983.10  
12          4750232680                    0.81           153649.65  
14          4750232680                    0.08           153649.65  
11          4750232680                    0.04           153649.65  
13          4750232680                    0.04           153649.65  
10          4750232680                    0.03           153649.65  
24          3218277220                    0.26            75044.36  
22          3218277220                    0.23            75044.36  
21          3218277220                    0.19            75044.36  
23          3218277220                    0.17            75044.36  
20          3218277220                    0.16            75044.36  
17            32519020                    0.56            49799.42  
19            32519020                    0.15            49799.42  
16            32519020                    0.11            4

In [64]:
for cluster in category_by_cluster['cluster_plus'].unique():
    # 해당 클러스터의 데이터 선택
    data = category_by_cluster[category_by_cluster['cluster_plus'] == cluster]
    
    # 막대그래프 생성
    fig = go.Figure(data=[go.Pie(labels=data['Category'], values=data['n_ratio_in_cluster'])])
    
    # 그래프 레이아웃 설정
    fig.update_layout(title=f'클러스터 {cluster} 구매 품목 비중',
                      xaxis_title='카테고리',
                      yaxis_title='Count')
    
    # 그래프 출력
    fig.show()

In [63]:
for cluster in category_by_cluster['cluster_plus'].unique():
    data = category_by_cluster[category_by_cluster['cluster_plus'] == cluster]
    
    fig = go.Figure(data=[go.Pie(labels=data['Category'], values=data['price_ratio_in_cluster'])])
    fig.update_layout(title= f'클러스터 {cluster} 매출 카테고리 비중')
    fig.show()